### PR TITLE
support waiting for things to happen in Flutter Driver

### PR DESCRIPTION
**Pull request by yjbanov**

### CODE REVIEW

File: packages/flutter_driver/lib/src/matcher_util.dart [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+/// Result of matching a value against a matcher.
+class MatchResult {
+  /// Whether the match succeeded.
+  final bool hasMatched;
+
+  /// If the match did not succeed, this field contains the explanation.
+  final String mismatchDescription;
+
+  MatchResult._matched()
```
Comment (devoncarew) [inline packages/flutter_driver/lib/src/matcher_util.dart]: You'll want to sort these ctors above the fields.

File: packages/flutter_driver/lib/src/retry.dart [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+/// that evaluates to the result.
+typedef dynamic Action();
+
+/// Performs [action] repeatedly until it either succeeds or [timeout] limit is
+/// reached.
+///
+/// When the retry time out, the last seen error and stack trace are returned in
+/// an error [Future].
+Future<dynamic> retry(Action action, Duration timeout,
+    Duration pauseBetweenRetries) async {
```
Comment (devoncarew) [inline packages/flutter_driver/lib/src/retry.dart]: nit: newline before `)`

File: packages/flutter_driver/test/flutter_driver_test.dart [hunk 3]
```diff
@@ -10,6 +10,7 @@ import 'package:flutter_driver/src/health.dart';
 import 'package:flutter_driver/src/message.dart';
 import 'package:json_rpc_2/json_rpc_2.dart' as rpc;
 import 'package:mockito/mockito.dart';
+import 'package:quiver/testing/async.dart';
```
Comment (devoncarew) [inline packages/flutter_driver/test/flutter_driver_test.dart]: Do we have an explicit dep on `quiver` in the pubspec?

File: packages/flutter_driver/lib/src/driver.dart [hunk 4]
```diff
@@ -4,20 +4,32 @@
 
 import 'dart:async';
 import 'package:vm_service_client/vm_service_client.dart';
+import 'package:matcher/matcher.dart';
```
Comment (devoncarew) [inline packages/flutter_driver/lib/src/driver.dart]: We should add `matcher` to the pubspec; I think it's being brought in now as a transitive dep.

**Comment by devoncarew** [comment]
lgtm! It'd be great to see the dartdoc for the public API of this package. I'm not sure which ones we're generating or hosting, other than the main flutter package (http://docs.flutter.io/flutter/).

File: packages/flutter_driver/lib/src/retry.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter_driver/lib/src/retry.dart]: nit nit: actually here the ( has no newline after it so the ) doesn't need a newline before it. :-) The rule is just that the whitespace after a (/{/[ should match the whitespace before the corresponding )/}/].

File: packages/flutter_driver/test/flutter_driver_test.dart (hunk 3 above)
Comment (yjbanov) [inline packages/flutter_driver/test/flutter_driver_test.dart]: Done. Also did a general clean-up of `pubspec.yaml` (removed obsolete dep on `logging`, sorted alphabetically).

File: packages/flutter_driver/lib/src/retry.dart (hunk 2 above)
Comment (yjbanov) [inline packages/flutter_driver/lib/src/retry.dart]: Tried with newlines on both sides. The current format looks better. Leaving as is.

File: packages/flutter_driver/lib/src/matcher_util.dart (hunk 1 above)
Comment (yjbanov) [inline packages/flutter_driver/lib/src/matcher_util.dart]: done

File: packages/flutter_driver/lib/src/driver.dart (hunk 4 above)
Comment (yjbanov) [inline packages/flutter_driver/lib/src/driver.dart]: done

**Comment by yjbanov** [comment]
@devoncarew 

> It'd be great to see the dartdoc for the public API of this package.

Yes! However, I'm not quite ready to promote the API yet. I'd like to dogfood it myself for a bit; get it to have a critical mass of functionality so it's actually useful. All I did so far is tab on a fab, verify some plain text. A far cry from testing a real app :)

**Pull request merged**